### ISSUE TITLE
many (trivial) warnings triggered by the header files

### ISSUE THREAD

**Issue by bmorel** [recastnavigation/recastnavigation#510]
Hello.
When building tools which include recastnavigation, I've noticed that there is a lot of trivial warnings in the headers, which could easily be fixed:

```
daemonmap/build% sed -n 's/^.*recast.*\.h.*-W\([^]]*\).*/\1/g p' log | sort | uniq -c | sort -n
      1 suggest-destructor-override
      5 suggest-override
      8 implicit-int-conversion
     20 weak-vtables
     32 extra-semi
     46 zero-as-null-pointer-constant
     82 sign-conversion
    253 old-style-cast
```

Many of those warnings are really trivial to fix (there are some others which can be get, but I disabled some which mostly do noise: no-padded, no-exit-time-destructors, no-global-constructors, no-missing-prototypes), but would a PR to fix them be accepted?
Concerning the warnings I have reported above:

* old-style-cast: those are most often trivial and helps understanding what kind of cast is needed, but they make the code longer (especially since I usually do some assertions depending on the type of cast, to help catch problems)
* sign conversion are usually because of the use of `int` for sizes
* zero-as-null-pointer-constant is mostly fixable with a sed expression
* extra-semi is, in recast's case, about having a `;` after a function's implementation
* weak vtables is mostly about slowing down compilation/linking, would just require to move at least one member function in a cpp file
* implicit int conversions is just about adding a cast (and usually I prefer to add an assert above just in case)
* overrides are just about style, but can help catch typos

I know this is not very important, and can trigger conflicts in current PRs which are likely more important, that's why I'm asking before doing the task. I'm merely asking for permission to do it, here (because I enjoy having things compiling without warnings).

Note that this is only the headers, which are emitting warnings when included. There are likely more of those in the cpp side, but those will be more likely to make conflicts, for possibly less value.

**Comment by grahamboree** [comment]
Appreciate you raising this and for providing the detailed information!  

We aim for Recast to be maximally compatible, and to that end we restrict the code to C++98.  Some of these warnings are related to features added in C++11 or later such as the `override` and `nullptr` keywords.  Though others appear to be issues agnostic to dialect, so they should be addressed.  

I'm surprised that there's just way more warnings enabled in clang for modern C++ than there are for C++98.  It's possible some of these arise from standard clarifications that were only made in modern C++ revisions.  🤔 Either way I agree we definitely should try to reduce the warnings noise here as much as possible.

**Comment by bmorel** [comment]
So, considering the use of C++98 (or C++03?), I guess the list of warnings to clean would be:

* implicit-int-conversion
* weak-vtables
* extra-semi
* sign-conversion
* old-style-cast

As I said, I would gladly do this, but no point at doing the job in double? I also happen to have much free time those days.

> I'm surprised that there's just way more warnings enabled in clang for modern C++

Well, I do have system using pre-nazi-flags for compiler by default: -Wall -Weverything. Pre-nazi, because no -Werror (otherwise nothing would compile, really, which is not the point) plus I removed some stuff about documentation and stuff being incompatible with C++98 (that I usually don't care about). So it's not nazi yet, but close enough.

IIRC there's much more warnings, which are less trivial to fix or reduce (the ones I listed are globally fixable with a regex or something like that), but I wanted to see how this issue would be treated before trying to bring the topic on warnings such as padding, which can't be completely removed, but can be reduced simply by grouping members in structs/classes.
I can do a more complete list if you want, and filter out the warnings that rely on C++11 features.
